source/windows/bcrypt_rsa.c

```diff
@@ -0,0 +1,458 @@
+    key_pair_impl->base.impl = key_pair_impl;
+    key_pair_impl->base.allocator = allocator;
+    aws_byte_buf_init_copy_from_cursor(&key_pair_impl->base.priv, allocator, key);
+
+    struct aws_der_decoder *decoder = aws_der_decoder_new(allocator, key);
+
+    if (!decoder) {
+        goto on_error;
+    }
+
+    struct aws_rsa_private_key_pkcs1 private_key_data;
+    AWS_ZERO_STRUCT(private_key_data);
+    if (aws_der_decoder_load_private_rsa_pkcs1(decoder, &private_key_data)) {
+        goto on_error;
+    }
+
+    /* Hard to predict final blob size, so use pkcs1 key size as upper bound. */
+    size_t total_buffer_size = key.len + sizeof(BCRYPT_RSAKEY_BLOB);
+
+    if (aws_byte_buf_init(&key_pair_impl->key_buf, allocator, total_buffer_size)) {
+        goto on_error;
+    }
+
+    aws_byte_buf_secure_zero(&key_pair_impl->key_buf);
+
+    BCRYPT_RSAKEY_BLOB key_blob;
+    AWS_ZERO_STRUCT(key_blob);
+    key_blob.Magic = BCRYPT_RSAFULLPRIVATE_MAGIC;
+    key_blob.BitLength = (ULONG)private_key_data.modulus.len * 8;
+    key_blob.cbPublicExp = (ULONG)private_key_data.publicExponent.len;
+    key_blob.cbModulus = (ULONG)private_key_data.modulus.len;
+    key_blob.cbPrime1 = (ULONG)private_key_data.prime1.len;
+    key_blob.cbPrime2 = (ULONG)private_key_data.prime2.len;
+
+    struct aws_byte_cursor header = aws_byte_cursor_from_array(&key_blob, sizeof(key_blob));
+    aws_byte_buf_append(&key_pair_impl->key_buf, &header);
+
+    LPCWSTR blob_type = BCRYPT_RSAFULLPRIVATE_BLOB;
+    ULONG flags = 0;
+
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.publicExponent);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.modulus);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.prime1);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.prime2);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.exponent1);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.exponent2);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.coefficient);
+    aws_byte_buf_append(&key_pair_impl->key_buf, &private_key_data.privateExponent);
+
+    NTSTATUS status = BCryptImportKeyPair(
+        s_rsa_alg,
+        NULL,
+        blob_type,
+        &key_pair_impl->key_handle,
+        key_pair_impl->key_buf.buffer,
+        (ULONG)key_pair_impl->key_buf.len,
+        flags);
+
+    if (!BCRYPT_SUCCESS(status)) {
+        aws_raise_error(AWS_ERROR_CAL_CRYPTO_OPERATION_FAILED);
+        goto on_error;
+    }
+
+    key_pair_impl->base.vtable = &s_rsa_key_pair_vtable;
+    key_pair_impl->base.key_size_in_bits = private_key_data.modulus.len * 8;
+
+    aws_der_decoder_destroy(decoder);
+
+    return &key_pair_impl->base;
+
+on_error:
+    aws_der_decoder_destroy(decoder);
+    aws_byte_buf_clean_up_secure(&key_pair_impl->base.priv);
+    aws_byte_buf_clean_up_secure(&key_pair_impl->base.pub);
+    s_rsa_destroy_key(&key_pair_impl->base);
```

Comment:
same feedback I've given elsewhere about you should probably just call 1 destroy function, instead of trying to destroy the individual parts like this